### DISCUSSION TITLE
E402 triggered after sys.path.append

### DISCUSSION THREAD

**Discussion by remchuk** [astral-sh/ruff#15133] [Q&A] [answered]
Hey there fellows,

I have quite a large python project with flask, with a main app (a directory called A with many directories inside).

In addition to A, I have several directories that are called `*_scripts` that usually call some functions from the main app( Directory A).

Is this intended? If feel like it is a false positive.

example:
```
A/
|
---- B/
     |  
     ------- likable.py
---- C/
----- D/
|
cool_scripts/
      |
      ----- nice_script.py

```

In nice_script.py I have the following:

```
#!/usr/bin/env python3

import os
import sys

ROOT_DIR = os.path.dirname(os.path.abpath(__file__))
sys.path.append(f"{ROOT_DIR}/../")

from A.B.likable import some_function
```

but  `from A.B.likable import some_function` prompts E402.

You are much wiser and more experiences people of python, any advice? is it false positive?
I found some similer issues but all of them doesn't seem to solve this issue.

**Comment by MichaReiser** [comment]
That Ruff triggers E402 on your example does make sense to me. E402 enforces that all imports are at the top of the file. This isn't the case here because of your path-manipulation statements

```
ROOT_DIR = os.path.dirname(os.path.abpath(__file__))
sys.path.append(f"{ROOT_DIR}/../")
```

Now, if it isn't important to you that imports are at the top, then I suggest to disable `E402` for all files in the scripts directory by using [`per-file-ignores`](https://docs.astral.sh/ruff/settings/#lint_per-file-ignores).

**Reply by remchuk** [reply]
It is important to me, but I'm unable to import the them before those lines are ran.

I found this pr:
https://github.com/astral-sh/ruff/pull/9047

Shouldn't it solve my problem?
(Its doesn't currently)

**Reply by MichaReiser** [reply, marked as answer]
Oh nice find. I wasn't aware of this behavior. I had a quick look at the implementation and I suspect the problem is that you have two statements. Can you try inlining `ROOT_DIR` like this

```py
#!/usr/bin/env python3

import os
import sys

sys.path.append(f"{os.path.dirname(os.path.abpath(__file__))}/../")

from A.B.likable import some_function
```

**Reply by remchuk** [reply]
I tried, but it isn't helping.

**Reply by remchuk** [reply]
> Oh nice find. I wasn't aware of this behavior. I had a quick look at the implementation and I suspect the problem is that you have two statements. Can you try inlining `ROOT_DIR` like this
> 
> ```python
> #!/usr/bin/env python3
> 
> import os
> import sys
> 
> sys.path.append(f"{os.path.dirname(os.path.abpath(__file__))}/../")
> 
> from A.B.likable import some_function
> ```

I even tried it with Path instead of the os.path.dirname and also to no avail.

**Reply by MichaReiser** [reply]
I'm unable to reproduce your issue. I see `E402` if I paste your original snippet but the error goes away if I paste the snippet from my reply in https://github.com/astral-sh/ruff/discussions/15133#discussioncomment-11683617 or the snippet that you just shared. Can you share your ruff configuration and how you call Ruff?

**Reply by remchuk** [reply]
pyproject.toml:
```
[tool.ruff]
line-length = 140

[tool.ruff.lint]
select = ["ALL"]
ignore = [
     "D",
     "FBT",
     "E501",
     "S311",
]

fixable = ["ALL"]

[tool.ruff.format]
line-ending = "auto"
indent-style = "space"
quota-style = "double"
skip-magic-trailing-comma = "false"

[tool.ruff.lint.isort]
combine-as-imports = true
force-sort-within-sections = true
split-on-trailing-comma = false

```

I'm using VSCode, so I have the extension installed and the module as well.

setting.json:

```
{
"ruff.lineLenght": 140,
"ruff.lint.select": [],
"ruff.lint.run": "onSave",
"ruff.configuration": "path/pyproject.toml"
"ruff.enable": true
"ruff.organizeImports": true,
"notebook.defaultFormatter": "charliermarsh.ruff"
"editor.defaultFormatter": "charliermarsh.ruff"
"ruff.configurationPreference": "filesystemFirst"
}
```

**Reply by MichaReiser** [reply]
The last two lines in your `format` configuration look incorrect. They should be:

```toml
quote-style = "double"
skip-magic-trailing-comma = false
```

Can you update the configuration and then try again?

**Reply by remchuk** [reply]
> The last two lines in your `format` configuration look incorrect. They should be:
> 
> ```toml
> quote-style = "double"
> skip-magic-trailing-comma = false
> ```
> 
> Can you update the configuration and then try again?

Oh sorry, I was copying it by hand due to working in air gapped environment.
They are correct in my conf file.

**Reply by MichaReiser** [reply]
It then must be something else with your setup because. 
Can you try running `ruff check --show-settings` and `ruff check --verbose` and share the output.

**Reply by remchuk** [reply]
it is quite a large file and it is quite difficult to paste all of it here.

Is there anything specific you want to see?

**Reply by MichaReiser** [reply]
Not really. I'm mainly trying to narrow down why it is working for me but not for you.

By the way, what Ruff version are you using?

**Reply by remchuk** [reply]
> Not really. I'm mainly trying to narrow down why it is working for me but not for you.
> 
> By the way, what Ruff version are you using?

0.7.3

**Reply by MichaReiser** [reply]
Thanks. I don't know what's "wrong" and I'm not sure how I can help you without more information but using the configuration you shared (after fixing the typos) and using Ruff 0.7.3, I get the following (expected) output:

```
❯ uvx ruff@0.7.3 check ./test.py
test.py:1:1: EXE001 Shebang is present but file is not executable
  |
1 | #!/usr/bin/env python3
  | ^^^^^^^^^^^^^^^^^^^^^^ EXE001
2 | 
3 | import os
  |

test.py:6:20: PTH120 `os.path.dirname()` should be replaced by `Path.parent`
  |
4 | import sys
5 | 
6 | sys.path.append(f"{os.path.dirname(os.path.abpath(__file__))}/../")
  |                    ^^^^^^^^^^^^^^^ PTH120
  |

test.py:9:25: F401 [*] `A.B.likable.some_function` imported but unused
  |
9 | from A.B.likable import some_function
  |                         ^^^^^^^^^^^^^ F401
  |
  = help: Remove unused import: `A.B.likable.some_function`

test.py:9:38: W292 [*] No newline at end of file
  |
9 | from A.B.likable import some_function
  |                                       W292
  |
  = help: Add trailing newline

Found 4 errors.
[*] 2 fixable with the `--fix` option.

~/astral/test via 🐍 v3.13.1 
```

and `test.py is

```py
#!/usr/bin/env python3

import os
import sys

sys.path.append(f"{os.path.dirname(os.path.abpath(__file__))}/../")


from A.B.likable import some_function
```

**Reply by remchuk** [reply]
You are definitely correct.
I totally missed one line that was present before the sys.append.
THANKS A LOT!

it works!
I'm sorry for the trouble!
